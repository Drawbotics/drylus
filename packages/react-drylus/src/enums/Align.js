import Enum from '@drawbotics/enums';


const Align = new Enum(
  'LEFT',
  'CENTER',
  'RIGHT',
);


export default Align;
