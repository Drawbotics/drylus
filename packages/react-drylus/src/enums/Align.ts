export enum Align {
  LEFT = 'LEFT',
  CENTER = 'CENTER',
  RIGHT = 'RIGHT',
  JUSTIFY = 'JUSTIFY',
}
