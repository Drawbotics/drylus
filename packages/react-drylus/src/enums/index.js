export Category from './Category';
export Tier from './Tier';
export Size from './Size';
export Position from './Position';
export Align from './Align';
export Color from './Color';
export Shade from './Shade';
