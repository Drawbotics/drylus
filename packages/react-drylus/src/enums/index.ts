export * from './Category';
export * from './Tier';
export * from './Size';
export * from './Position';
export * from './Align';
export * from './Color';
export * from './Shade';
