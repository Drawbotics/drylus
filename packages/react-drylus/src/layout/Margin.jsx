import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import sv from '@drawbotics/style-vars';
import camelCase from 'lodash/camelCase';

import Sizes from '../base/Sizes';


const styles = {
  root: css`
    margin: ${sv.defaultMargin};
  `,
  resetMargin: css`
    margin: 0;
  `,
  extraSmall: css`
    margin: ${sv.marginExtraSmall};
  `,
  small: css`
    margin: ${sv.marginSmall};
  `,
  large: css`
    margin: ${sv.marginLarge};
  `,
  extraLarge: css`
    margin: ${sv.marginExtraLarge};
  `,
  extraSmallLeft: css`
    margin-left: ${sv.marginExtraSmall};
  `,
  extraSmallRight: css`
    margin-right: ${sv.marginExtraSmall};
  `,
  extraSmallTop: css`
    margin-top: ${sv.marginExtraSmall};
  `,
  extraSmallBottom: css`
    margin-bottom: ${sv.marginExtraSmall};
  `,
  smallLeft: css`
    margin-left: ${sv.marginSmall};
  `,
  smallRight: css`
    margin-right: ${sv.marginSmall};
  `,
  smallTop: css`
    margin-top: ${sv.marginSmall};
  `,
  smallBottom: css`
    margin-bottom: ${sv.marginSmall};
  `,
  defaultLeft: css`
    margin-left: ${sv.defaultMargin};
  `,
  defaultRight: css`
    margin-right: ${sv.defaultMargin};
  `,
  defaultTop: css`
    margin-top: ${sv.defaultMargin};
  `,
  defaultBottom: css`
    margin-bottom: ${sv.defaultMargin};
  `,
  largeLeft: css`
    margin-left: ${sv.marginLarge};
  `,
  largeRight: css`
    margin-right: ${sv.marginLarge};
  `,
  largeTop: css`
    margin-top: ${sv.marginLarge};
  `,
  largeBottom: css`
    margin-bottom: ${sv.marginLarge};
  `,
  extraLargeLeft: css`
    margin-left: ${sv.marginExtraLarge};
  `,
  extraLargeRight: css`
    margin-right: ${sv.marginExtraLarge};
  `,
  extraLargeTop: css`
    margin-top: ${sv.marginExtraLarge};
  `,
  extraLargeBottom: css`
    margin-bottom: ${sv.marginExtraLarge};
  `,
};


const Margin = ({
  children,
  size,
  style,
}) => {
  const isUniform = typeof size !== 'object';
  return (
    <div className={cx(styles.root, {
      [styles[camelCase(size)]]: isUniform && size,
      [styles.resetMargin]: ! isUniform,
      [styles[camelCase(`${size.left}_LEFT`)]]: ! isUniform && size.left,
      [styles[camelCase(`${size.right}_RIGHT`)]]: ! isUniform && size.right,
      [styles[camelCase(`${size.top}_TOP`)]]: ! isUniform && size.top,
      [styles[camelCase(`${size.bottom}_BOTTOM`)]]: ! isUniform && size.bottom,
    })} style={style}>
      {children}
    </div>
  );
};


Margin.propTypes = {
  /** Determines the amount of margin given to the component. If a single value, the margin is applied equally to each side */
  size: PropTypes.oneOfType([
    PropTypes.oneOf([ Sizes.DEFAULT, Sizes.SMALL, Sizes.EXTRA_SMALL, Sizes.LARGE, Sizes.EXTRA_LARGE ]),
    PropTypes.shape({
      left: PropTypes.oneOf([ Sizes.DEFAULT, Sizes.SMALL, Sizes.EXTRA_SMALL, Sizes.LARGE, Sizes.EXTRA_LARGE ]),
      right: PropTypes.oneOf([ Sizes.DEFAULT, Sizes.SMALL, Sizes.EXTRA_SMALL, Sizes.LARGE, Sizes.EXTRA_LARGE ]),
      bottom: PropTypes.oneOf([ Sizes.DEFAULT, Sizes.SMALL, Sizes.EXTRA_SMALL, Sizes.LARGE, Sizes.EXTRA_LARGE ]),
      top: PropTypes.oneOf([ Sizes.DEFAULT, Sizes.SMALL, Sizes.EXTRA_SMALL, Sizes.LARGE, Sizes.EXTRA_LARGE ]),
    }),
  ]),

  /** The content of the margin wrapper */
  children: PropTypes.node,

  /** Prop to override any style if necessary, use sparingly */
  style: PropTypes.object,
};


export default Margin;
