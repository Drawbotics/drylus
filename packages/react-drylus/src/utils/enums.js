import camelCase from 'lodash/camelCase';

import { Category, Color, Shade, Tier } from '../enums';

export function getEnumAsClass(enumVal) {
  return camelCase(enumVal?.description?.toLowerCase());
}

export function colorEnumToCategory(enumVal) {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
  return enumVal;
}

export function categoryEnumToColor(enumVal) {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
  return enumVal;
}

export function shadeEnumToTier(enumVal) {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
  return enumVal;
}

export function getIconForCategory(category) {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
