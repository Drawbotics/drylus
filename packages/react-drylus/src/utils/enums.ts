import camelCase from 'lodash/camelCase';

import { Category, Color, Position, Shade, Size, Tier } from '../enums';

type Enum = Category | Color | Shade | Tier | Size | Position;

export function getEnumAsClass<T extends Record<string, string>>(enumVal?: Enum): keyof T {
  return camelCase(enumVal?.toLowerCase());
}

export function colorEnumToCategory(enumVal: Color): Category | undefined {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
}

export function categoryEnumToColor(enumVal: Category): Color | undefined {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
}

export function shadeEnumToTier(enumVal: Shade): Tier | undefined {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
}

export function getIconForCategory(category: Category): string {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
