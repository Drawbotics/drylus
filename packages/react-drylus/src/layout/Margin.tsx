import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import camelCase from 'lodash/camelCase';
import React from 'react';

import { Size } from '../enums';
import { Rectangular, Responsive, Style, Variable } from '../types';
import { useResponsiveProps } from '../utils';

const styles = {
  root: css`
    margin: ${sv.defaultMargin};

    @media ${sv.screenL} {
      margin: ${sv.marginSmall};
    }
  `,
  resetMargin: css`
    margin: 0;

    @media ${sv.screenL} {
      margin: 0;
    }
  `,
  extraSmall: css`
    margin: ${sv.marginExtraSmall};

    @media ${sv.screenL} {
      margin: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  small: css`
    margin: ${sv.marginSmall};

    @media ${sv.screenL} {
      margin: ${sv.marginExtraSmall};
    }
  `,
  large: css`
    margin: ${sv.marginLarge};

    @media ${sv.screenL} {
      margin: ${sv.defaultMargin};
    }
  `,
  extraLarge: css`
    margin: ${sv.marginExtraLarge};

    @media ${sv.screenL} {
      margin: ${sv.marginLarge};
    }
  `,
  huge: css`
    margin: ${sv.marginHuge};

    @media ${sv.screenL} {
      margin: ${sv.marginExtraLarge};
    }
  `,
  extraHuge: css`
    margin: ${sv.marginExtraHuge};

    @media ${sv.screenL} {
      margin: ${sv.marginHuge};
    }
  `,
  massive: css`
    margin: ${sv.marginMassive};

    @media ${sv.screenL} {
      margin: ${sv.marginExtraHuge};
    }
  `,
  extraSmallLeft: css`
    margin-left: ${sv.marginExtraSmall};

    @media ${sv.screenL} {
      margin-left: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  extraSmallRight: css`
    margin-right: ${sv.marginExtraSmall};

    @media ${sv.screenL} {
      margin-right: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  extraSmallTop: css`
    margin-top: ${sv.marginExtraSmall};

    @media ${sv.screenL} {
      margin-top: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  extraSmallBottom: css`
    margin-bottom: ${sv.marginExtraSmall};

    @media ${sv.screenL} {
      margin-bottom: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  smallLeft: css`
    margin-left: ${sv.marginSmall};

    @media ${sv.screenL} {
      margin-left: ${sv.marginExtraSmall};
    }
  `,
  smallRight: css`
    margin-right: ${sv.marginSmall};

    @media ${sv.screenL} {
      margin-right: ${sv.marginExtraSmall};
    }
  `,
  smallTop: css`
    margin-top: ${sv.marginSmall};

    @media ${sv.screenL} {
      margin-top: ${sv.marginExtraSmall};
    }
  `,
  smallBottom: css`
    margin-bottom: ${sv.marginSmall};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginExtraSmall};
    }
  `,
  defaultLeft: css`
    margin-left: ${sv.defaultMargin};

    @media ${sv.screenL} {
      margin-left: ${sv.marginSmall};
    }
  `,
  defaultRight: css`
    margin-right: ${sv.defaultMargin};

    @media ${sv.screenL} {
      margin-right: ${sv.marginSmall};
    }
  `,
  defaultTop: css`
    margin-top: ${sv.defaultMargin};

    @media ${sv.screenL} {
      margin-top: ${sv.marginSmall};
    }
  `,
  defaultBottom: css`
    margin-bottom: ${sv.defaultMargin};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginSmall};
    }
  `,
  largeLeft: css`
    margin-left: ${sv.marginLarge};

    @media ${sv.screenL} {
      margin-left: ${sv.defaultMargin};
    }
  `,
  largeRight: css`
    margin-right: ${sv.marginLarge};

    @media ${sv.screenL} {
      margin-right: ${sv.defaultMargin};
    }
  `,
  largeTop: css`
    margin-top: ${sv.marginLarge};

    @media ${sv.screenL} {
      margin-top: ${sv.defaultMargin};
    }
  `,
  largeBottom: css`
    margin-bottom: ${sv.marginLarge};

    @media ${sv.screenL} {
      margin-bottom: ${sv.defaultMargin};
    }
  `,
  extraLargeLeft: css`
    margin-left: ${sv.marginExtraLarge};

    @media ${sv.screenL} {
      margin-left: ${sv.marginLarge};
    }
  `,
  extraLargeRight: css`
    margin-right: ${sv.marginExtraLarge};

    @media ${sv.screenL} {
      margin-right: ${sv.marginLarge};
    }
  `,
  extraLargeTop: css`
    margin-top: ${sv.marginExtraLarge};

    @media ${sv.screenL} {
      margin-top: ${sv.marginLarge};
    }
  `,
  extraLargeBottom: css`
    margin-bottom: ${sv.marginExtraLarge};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginLarge};
    }
  `,
  hugeLeft: css`
    margin-left: ${sv.marginHuge};

    @media ${sv.screenL} {
      margin-left: ${sv.marginExtraLarge};
    }
  `,
  hugeRight: css`
    margin-right: ${sv.marginHuge};

    @media ${sv.screenL} {
      margin-right: ${sv.marginExtraLarge};
    }
  `,
  hugeTop: css`
    margin-top: ${sv.marginHuge};

    @media ${sv.screenL} {
      margin-top: ${sv.marginExtraLarge};
    }
  `,
  hugeBottom: css`
    margin-bottom: ${sv.marginHuge};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginExtraLarge};
    }
  `,
  extraHugeLeft: css`
    margin-left: ${sv.marginExtraHuge};

    @media ${sv.screenL} {
      margin-left: ${sv.marginHuge};
    }
  `,
  extraHugeRight: css`
    margin-right: ${sv.marginExtraHuge};

    @media ${sv.screenL} {
      margin-right: ${sv.marginHuge};
    }
  `,
  extraHugeTop: css`
    margin-top: ${sv.marginExtraHuge};

    @media ${sv.screenL} {
      margin-top: ${sv.marginHuge};
    }
  `,
  extraHugeBottom: css`
    margin-bottom: ${sv.marginExtraHuge};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginHuge};
    }
  `,
  massiveLeft: css`
    margin-left: ${sv.marginMassive};

    @media ${sv.screenL} {
      margin-left: ${sv.marginExtraHuge};
    }
  `,
  massiveRight: css`
    margin-right: ${sv.marginMassive};

    @media ${sv.screenL} {
      margin-right: ${sv.marginExtraHuge};
    }
  `,
  massiveTop: css`
    margin-top: ${sv.marginMassive};

    @media ${sv.screenL} {
      margin-top: ${sv.marginExtraHuge};
    }
  `,
  massiveBottom: css`
    margin-bottom: ${sv.marginMassive};

    @media ${sv.screenL} {
      margin-bottom: ${sv.marginExtraHuge};
    }
  `,
};

function _computeSize(sizeDescription: Rectangular) {
  let size: Variable = {};

  if (sizeDescription.horizontal) {
    size.left = sizeDescription.horizontal;
    size.right = sizeDescription.horizontal;
  }
  if (sizeDescription.vertical) {
    size.top = sizeDescription.vertical;
    size.bottom = sizeDescription.vertical;
  }

  return size;
}

function _isRectangular(size: any): size is Rectangular {
  return size.vertical || size.horizontal;
}

export interface MarginProps {
  /** Determines the amount of margin given to the component. If a single value, the margin is applied equally to each side */
  size?: Size | Rectangular | Variable;

  /** The content of the margin wrapper */
  children?: React.ReactNode;

  /** Prop to override any style if necessary, use sparingly */
  style?: Style;

  /** Used for style overrides */
  className?: string;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;
}

export const Margin = ({ responsive, ...rest }: MarginProps) => {
  const { children, size: rawSize, style, className } = useResponsiveProps<MarginProps>(
    rest,
    responsive,
  );

  const isUniform = typeof rawSize !== 'object';

  const size = !isUniform && _isRectangular(rawSize) ? _computeSize(rawSize) : rawSize;
  return (
    <div
      className={cx(
        styles.root,
        {
          [styles[camelCase(size as Size) as keyof typeof styles]]: isUniform && size != null,
          [styles.resetMargin]: !isUniform,
          [styles[camelCase(`${(size as Variable)?.left}_LEFT`) as keyof typeof styles]]:
            !isUniform && (size as Variable)?.left != null,
          [styles[camelCase(`${(size as Variable)?.right}_RIGHT`) as keyof typeof styles]]:
            !isUniform && (size as Variable)?.right != null,
          [styles[camelCase(`${(size as Variable)?.top}_TOP`) as keyof typeof styles]]:
            !isUniform && (size as Variable)?.top != null,
          [styles[camelCase(`${(size as Variable)?.bottom}_BOTTOM`) as keyof typeof styles]]:
            !isUniform && (size as Variable)?.bottom != null,
        },
        className,
      )}
      style={style}>
      {children}
    </div>
  );
};
