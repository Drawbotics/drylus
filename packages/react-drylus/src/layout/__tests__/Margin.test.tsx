import React from 'react';
import { create } from 'react-test-renderer';

import { Size } from '../../enums';
import { Margin } from '../Margin';

describe('Margin', () => {
  describe('matches snapshot when', () => {
    it('has a uniform size', () => {
      const tree = create(<Margin size={Size.LARGE}>Content</Margin>).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a size only on 1 side', () => {
      const tree = create(<Margin size={{ left: Size.LARGE }}>Content</Margin>).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a different size on each side', () => {
      const tree = create(
        <Margin
          size={{
            left: Size.LARGE,
            top: Size.SMALL,
            bottom: Size.DEFAULT,
            right: Size.EXTRA_SMALL,
          }}>
          Content
        </Margin>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has vertical and horizontal sizes', () => {
      const tree = create(
        <Margin
          size={{
            vertical: Size.LARGE,
            horizontal: Size.SMALL,
          }}>
          Content
        </Margin>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
