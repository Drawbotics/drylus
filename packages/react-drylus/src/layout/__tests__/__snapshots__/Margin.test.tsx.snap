// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Margin matches snapshot when has a different size on each side 1`] = `
.emotion-0 {
  margin: 24px;
  margin: 0;
  margin-left: 32px;
  margin-right: 8px;
  margin-top: 16px;
  margin-bottom: 24px;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 16px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 0;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-left: 24px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-right: calc(8px / 2);
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-top: 8px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-bottom: 16px;
  }
}

<div
  className="emotion-0"
>
  Content
</div>
`;

exports[`Margin matches snapshot when has a size only on 1 side 1`] = `
.emotion-0 {
  margin: 24px;
  margin: 0;
  margin-left: 32px;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 16px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 0;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-left: 24px;
  }
}

<div
  className="emotion-0"
>
  Content
</div>
`;

exports[`Margin matches snapshot when has a uniform size 1`] = `
.emotion-0 {
  margin: 24px;
  margin: 32px;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 16px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 24px;
  }
}

<div
  className="emotion-0"
>
  Content
</div>
`;

exports[`Margin matches snapshot when has vertical and horizontal sizes 1`] = `
.emotion-0 {
  margin: 24px;
  margin: 0;
  margin-left: 16px;
  margin-right: 16px;
  margin-top: 32px;
  margin-bottom: 32px;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 16px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin: 0;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-left: 8px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-right: 8px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-top: 24px;
  }
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    margin-bottom: 24px;
  }
}

<div
  className="emotion-0"
>
  Content
</div>
`;
